graphene-writer/src/main/kotlin/com/graphene/writer/store/key/AbstractElasticsearchKeyStoreHandler.kt

```diff
@@ -0,0 +1,160 @@
+
+  private lateinit var client: TransportClient
+  private lateinit var scheduler: ScheduledExecutorService
+  private lateinit var request: MetricMultiGetRequestBuilder
+  private lateinit var index: String
+  private lateinit var type: String
+  private lateinit var bulkProcessor: BulkProcessor
+  private lateinit var grapheneKeyMapper: GrapheneKeyMapper
+
+  private var lastFlushTimeSeconds = DateTimeUtils.currentTimeSeconds()
+  private var batchSize: Int = 0
+  private var flushInterval: Long = 0
+
+  private val metrics = ConcurrentLinkedQueue<GrapheneMetric>()
+
+  @PostConstruct
+  fun init() {
+    client = elasticsearchFactory.transportClient()
+
+    bulkProcessor = elasticsearchFactory.bulkProcessor(client)
+    grapheneKeyMapper = GrapheneKeyMapper()
+
+    index = property.index
+    type = property.type
+    request = MetricMultiGetRequestBuilder(client, index, type)
+    batchSize = property.bulk!!.actions
+    flushInterval = property.bulk!!.interval
+
+    scheduler = Executors.newSingleThreadScheduledExecutor(NamedThreadFactory(SimpleKeyStoreHandler::class.simpleName!!))
+    scheduler.scheduleWithFixedDelay(this, 3000, 100, TimeUnit.MILLISECONDS)
+  }
+
+  override fun handle(grapheneMetric: GrapheneMetric) {
+    metrics.offer(grapheneMetric)
+  }
+
+  override fun run() {
+    try {
+      val metric = metrics.poll()
+      if (Objects.nonNull(metric)) {
+        addToBatch(metric)
+      }
+    } catch (e: Exception) {
+      logger.error("Encountered error in busy loop: ", e)
+    }
+
+    if (0 < request.size()) {
+      flush()
+    }
+  }
+
+  private fun addToBatch(metric: GrapheneMetric) {
+    request.add(metric)
+
+    if (batchSize <= request.size() || lastFlushTimeSeconds < DateTimeUtils.currentTimeSeconds() - flushInterval) {
+      flush()
+    }
+  }
+
+  private fun flush() {
+    val multiGetItemResponse = request.execute().actionGet()
+
+    for (response in multiGetItemResponse.responses) {
+      if (response.isFailed) {
+        logger.error("Get failed: " + response.failure.message)
+      }
+
+      val metric = request.metrics[response.id]
+      if (response.isFailed || !response.response.isExists) {
+        val parts = metric!!.getGraphiteKeyParts()
+        val graphiteKeySb = StringBuilder()
+
+        for (depth in parts.indices) {
+          if (graphiteKeySb.toString().isNotEmpty()) {
+            graphiteKeySb.append(".")
+          }
+          graphiteKeySb.append(parts[depth])
+          try {
+            val graphiteKeyPart = graphiteKeySb.toString()
+            bulkProcessor.add(IndexRequest(index, type, metric.getTenant() + "_" + graphiteKeyPart)
+              .source(source(metric.getTenant(), graphiteKeyPart, depth, isLeaf(depth, parts))))
+          } catch (e: IOException) {
+            logger.error(e)
+          }
+        }
+      }
+    }
+
+    request = MetricMultiGetRequestBuilder(client, index, type)
```

Comment:
Do we need this line?

Reply:
Without this logic, the request would not be initialized.
The request is to accumulate the incoming metrics and check if the key is already stored in ES. If not initialized, existing keys and new keys will continue to accumulate, causing problems such as OOM.